chapters/high_level_shader_language_comparison.adoc

```diff
@@ -0,0 +1,972 @@
+====
+
+=== Examples
+
+Writing positions from the vertex shader:
+
+GLSL:
+[source,glsl]
+----
+layout (location = 0) in vec4 inPos;
+
+void main() {
+    // The vertex output position is written to the gl_Position built-in
+    gl_Position = ubo.projectionMatrix * ubo.viewMatrix * ubo.modelMatrix * inPos.xyz;
+}
+----
+
+HLSL
+[source,hlsl]
+----
+struct VSOutput
+{
+    // The SV_POSITION semantic declares the Pos member as the vertex output position 
+	float4 Pos : SV_POSITION;
+};
+
+VSOutput main(VSInput input)
+{
+	VSOutput output = (VSOutput)0;
+	output.Pos = mul(ubo.projectionMatrix, mul(ubo.viewMatrix, mul(ubo.modelMatrix, input.Pos)));
+	return output;
+}
+----
+
+Reading the vertex index:
+
+GLSL:
+[source,glsl]
+----
+void main() 
+{
+    // The vertex index is stored in the gl_VertexIndex built-in
+    outUV = vec2((gl_VertexIndex << 1) & 2, gl_VertexIndex & 2);
+}
+----
+
+HLSL
+[source,hlsl]
+----
+struct VSInput
+{
+    // The SV_VertexID semantic declares the VertexIndex member as the vertex index input
+    uint VertexIndex : SV_VertexID
+};
+
+VSOutput main(VSInput input)
+{
+	VSOutput output = (VSOutput)0;
+	output.UV = float2((input.VertexIndex << 1) & 2, input.VertexIndex & 2);
+	return output;
+}
+----
+
+== Shader interface
+
+[NOTE]
+====
+Shader interfaces greatly differ between GLSL and HLSL. Vulkan concepts not directly available in HLSL use the link:https://github.com/microsoft/DirectXShaderCompiler/blob/main/docs/SPIR-V.rst#the-implicit-vk-namespace)[implicit vk namespace].
+====
+
+=== Descriptor bindings
+
+==== GLSL
+
+[source,glsl]
+----
+layout (set = <set-index>, binding = <binding-index>) uniform <type> <name>
+----
+
+There are two options for defining descriptor set and binding indices in HLSL when using Vulkan.
+
+==== HLSL way
+
+[source,hlsl]
+----
+<type> <name> : register(<register-type><binding-index>, space<set-index>)
+----
+
+Using this syntax, descriptor set and binding indices will be implicitly assigned from the set and binding index.
+
+==== Vulkan namespace
+
+[source,hlsl]
+----
+[[vk::binding(binding-index, set-index)]]
+<type> <name>
+----
+
+With this option, descriptor set and binding indices are explicitly set using `vk::binding`.
+
+[NOTE]
+====
+It's possible to use both the `vk::binding[]` and `register()` syntax for one descriptor. This can be useful if a shader is used for both Vulkan and DirectX.
+====
+
+==== Examples
+
+===== GLSL
+
+[source, glsl]
+----
+layout (set = 1, binding = 0) uniform Node {
+	mat4 matrix;
+} node;
+----
+
+===== HLSL
+----
+struct Node {
+	float4x4 transform;
+};
+
+// HLSL style
+cbuffer NodeBuf : register(b0, space1) { Node node; }
```

Comment:
This is legacy style. ConstantBuffer<T> is the modern way (and supports arrayed descriptors).